lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c

```diff
@@ -57,6 +57,70 @@ const ziti_tunnel_ctrl* ziti_tunnel_init_cmd(uv_loop_t *loop, tunneler_context t
     return &CMD_CTX.ctrl;
 }
 
+
+int ziti_dump_to_log_cb(void* stringsBuilder, const char *fmt,  ...) {
+    static char line[4096];
+
+    va_list vargs;
+    va_start(vargs, fmt);
+    vsnprintf(line, sizeof(line), fmt, vargs);
+    va_end(vargs);
+
+    if (strlen(stringsBuilder) + strlen(line) > 8192) {
+        return -1;
+    }
+    // write/append to the buffer
+    strncat(stringsBuilder, line, sizeof(line));
+    return 0;
+}
+
+void ziti_dump_to_log(void *ctx) {
+    char* buffer;
+    buffer = malloc(8192*sizeof(char));
+    buffer[0] = 0;
+    //actually invoke ziti_dump here
+    ziti_dump(ctx, ziti_dump_to_log_cb, buffer);
+    ZITI_LOG(INFO, "ziti dump to log %s", buffer);
+    free(buffer);
+}
+
+int ziti_dump_to_file_cb(void* fp, const char *fmt,  ...) {
+    static char line[4096];
+
+    va_list vargs;
+    va_start(vargs, fmt);
+    vsnprintf(line, sizeof(line), fmt, vargs);
```

Comment:
I think you could omit the fputs (and the temp buffer) and write to fp directly via vfprintf:

    vfprintf(fp, fmt, vargs);